operator: Move CRD registration to operator

christarazi opened the pull request:
## Brief description

This PR is the final piece to move the CRD registration responsibilities to cilium-operator. In the case of new CRDs, an upcoming PR will install the CRDs via Helm. In the case of upgrades, cilium-operator will handle the upgrades if Helm does not upgrade them. This will ensure that we are always running the most up-to-date CRDs.

See commit msgs.

Updates: https://github.com/cilium/cilium/issues/12737

```release-note
Cilium Operator will now handle CRD operations on behalf of Cilium Agent
```

christarazi committed:
operator: Move K8s init logic to function

This commit is mostly a refactoring change to ease future commits. It
also removes a duplicated `Update()` call which is already done in
`Init()`.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
operator: Move CRD registration to operator

In pursuit of delegating all cluster operations to cilium-operator from
the agent, this commit moves the CRD registration under the purview of
cilium-operator.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on operator/main.go:
```diff
@@ -177,34 +202,20 @@ func runOperator() {
 			k8sversion.Version(), k8sversion.MinimalVersionConstraint)
 	}
 
+	// Register the CRDs after validating that we are running on a supported
+	// version of K8s.
+	if err := k8s.RegisterCRDs(); err != nil {
+		log.WithError(err).Fatal("Unable to register CRDs")
+	}
```
One question I have is: should this be done only by the leader in case of Operation HA? Or is it safe for multiple Operators to try? 

My impression that it's likely OK because we will only modify the schema version to the latest, so if there was ever a conflicting change, the latest version will win out.

aanm replied:
Yes, to install the CRD faster it can be done outside of the leader election.

christarazi committed:
install: Grant operator permissions for CRD ops

This commit deprecates the permissions for CRD operations and marks them
for removal in v1.10 from the agent. Additionally, this commit grants
them to cilium-operator. This commit is a preparatory for a future
commit to move the CRD operations to the cilium-operator.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml:
```diff
@@ -55,11 +55,7 @@ rules:
   - get
-  - list
-  - watch
-  - update
```
This can't be removed, it can only be deprecated. Mark this for removal in 1.10. See 750c7bc3040b858f30af2b1957b30bdd5ad70c56 for more info.

aanm commented on install/kubernetes/cilium/charts/preflight/templates/clusterrole.yaml:
```diff
@@ -55,11 +55,7 @@ rules:
   - get
-  - list
-  - watch
-  - update
```
This can't be removed, it can only be deprecated. Mark this for removal in 1.10. See 750c7bc3040b858f30af2b1957b30bdd5ad70c56 for more info.

aanm reviewed (changes requested):
One other item missing. There is, all over the operator code, check methods waiting for a CRD to be available. This should no longer be required and can be removed as part of these changes.

fristonio reviewed (approved):
Other than the changes requested by @aanm LGTM :rocket:

sayboras commented on operator/main.go:
```diff
@@ -124,6 +124,31 @@ func initEnv() {
+	// in non HA mode. As this can conflict with other operator instances running
+	// in the cluster, that figured the capabilities correctly and is running in HA
+	// mode.
+	if err := k8sversion.Update(k8s.Client(), option.Config); err != nil {
```
is this step considered as part of init k8s ? If yes, shall we move close(k8sInitDone) only after this step. If no, we can move this step back to the main flow.

christarazi replied:
Thanks for pointing this out. See https://github.com/cilium/cilium/pull/13285#discussion_r495170897 for more details.

vadorovsky commented on operator/main.go:
```diff
@@ -124,6 +124,31 @@ func initEnv() {
+	// in non HA mode. As this can conflict with other operator instances running
+	// in the cluster, that figured the capabilities correctly and is running in HA
+	// mode.
+	if err := k8sversion.Update(k8s.Client(), option.Config); err != nil {
```
Isn't this done by `k8s.Init()` already? I think that in case we cannot update k8s version and capabilities, we will Fatal earlier in line 135.

christarazi replied:
Thanks for pointing this out. According to git blame, that block of code was added by @fristonio. I am going to remove it for now, and await his feedback.

fristonio replied:
Good catch, the changes look good. I think I missed that we already update capabilities in `k8s.Init`. Thanks @mrostecki

christarazi committed:
operator: Revert waiting for CRDs

This commit reverts the following two commits, along with the vendored
changes that are no longer needed. The rationale is that we no longer
need cilium-operator to wait for CRDs as cilium-operator will now
register the CRDs (in a future commit), previously done by the agent.

---

Commit 5dbe4133a92644278406a47df7892f6e8e7b44c2 ("operator: Make CRD
availability timeout configurable") was selectively reverted, only
keeping the declarations of the crd-wait-timeout flag, so that they can
be deprecated in a future commit.

---

Revert "operator: Wait for CRDs before running informers"

This reverts commit 8e4f348e1982a081547e90d32221aa058b6fb71c.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on Documentation/cmdref/cilium-operator-aws.md:
```diff
@@ -28,7 +28,6 @@ cilium-operator-aws [flags]
       --cnp-status-update-interval duration       Interval between CNP status updates sent to the k8s-apiserver per-CNP (default 1s)
       --config string                             Configuration file (default "$HOME/ciliumd.yaml")
       --config-dir string                         Configuration directory that contains a file for each option
-      --crd-wait-timeout duration                 Operator will exit if CRDs are not available within this duration upon startup (default 5m0s)
```
Do we need to deprecate these or can we just remove them for 1.9? If we can remove, then we'll need to update the docs too. @aanm

aanm commented on install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml:
```diff
@@ -55,11 +55,13 @@ rules:
   - list
   - watch
   - update
+
```
leftover?

christarazi replied:
No, I separated them to make it clearer which actions are deprecated.

aanm commented on install/kubernetes/cilium/charts/preflight/templates/clusterrole.yaml:
```diff
@@ -55,11 +55,13 @@ rules:
   - list
   - watch
   - update
+
```
leftover?

christarazi replied:
Same as above

aanm commented on install/kubernetes/cilium/charts/config/templates/configmap.yaml:
```diff
@@ -643,3 +636,8 @@ data:
+{{- if hasKey .Values "blacklistConflictingRoutes" }}
+  # Configure blacklisting of local routes not owned by Cilium.
+  blacklist-conflicting-routes: {{ .Values.blacklistConflictingRoutes | quote }}
+{{- end }}
```
this should not have been added

christarazi replied:
Thanks, removed.

aanm commented on operator/flags.go:
```diff
@@ -267,9 +267,6 @@ func init() {
 
-	flags.Duration(operatorOption.CRDWaitTimeout, 5*time.Minute, "Operator will exit if CRDs are not available within this duration upon startup")
-	option.BindEnv(operatorOption.CRDWaitTimeout)
-
```
This can't be removed, needs to be deprecated and mark for removal in 1.10 as the upstream PR was backported to 1.8

christarazi replied:
Added them back and added a commit to deprecate them separately.

christarazi committed:
operator: Deprecate crd-wait-timeout

This commit deprecates the crd-wait-timeout option as the functionality
has been removed in the previous commit. This option will be removed in
1.10. This commit also updates the cmdref accordingly.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

sayboras reviewed (approved):
LGTM :+1:

aanm commented on install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml:
```diff
@@ -55,11 +55,13 @@ rules:
   - list
   - watch
   - update
+
```
Please add a comment here stating that the "why" we need this get, which we only need it for the preflight so ideally we should really remove it from the agent.

aanm reviewed (approved):
One small TODO doesn't require my follow up review

fristonio reviewed (approved):
LGTM :rocket:
